src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Node/JsonArrayConverter.cs

```diff
@@ -34,10 +34,29 @@ public override void Write(Utf8JsonWriter writer, JsonArray? value, JsonSerializ
             }
         }
 
-        public static JsonArray ReadList(ref Utf8JsonReader reader, JsonNodeOptions? options = null)
+        public static JsonArray ReadList(ref Utf8JsonReader reader, JsonSerializerOptions  options)
         {
-            JsonElement jElement = JsonElement.ParseValue(ref reader);
-            return new JsonArray(jElement, options);
+            JsonNodeOptions nodeOptions = options.GetNodeOptions();
+
+            if (options.AllowDuplicateProperties)
+            {
+                JsonElement jElement = JsonElement.ParseValue(ref reader);
+                return new JsonArray(jElement, nodeOptions);
+            }
+            else if (options.PropertyNameCaseInsensitive)
+            {
+                // Do duplicate detection by expanding the JsonArray eagerly
+                JsonElement jElement = JsonElement.ParseValue(ref reader);
+                JsonArray jArray = new JsonArray(jElement, nodeOptions);
+                jArray.InitializeComplexNodesEagerly();
+                return jArray;
+            }
+            else
+            {
+                // Do duplicate detection with JsonElement.ParseValue
+                JsonElement jElement = JsonElement.ParseValue(ref reader, allowDuplicateProperties: false);
```

Comment:
Would be very interested to see benchmarks showing the performance of each of these branches.

Reply:
I removed this path and moved to just use the JsonNode directly like case-insensitive.